packages/protocol/contracts/governance/DowntimeSlasherIntervals.sol

```diff
@@ -0,0 +1,307 @@
+   * @param endBlocks List of blocks that ends a previously validated interval.
+   * @param signerIndices Indices of the signers within the validator set for every epoch change.
+   * @return True if the validator signature does not appear in any block within the window.
+   */
+  function wasDownForIntervals(
+    uint256[] memory startBlocks,
+    uint256[] memory endBlocks,
+    uint256[] memory signerIndices
+  ) public view returns (bool) {
+    require(startBlocks.length > 0, "requires at least one interval");
+    require(
+      startBlocks.length == endBlocks.length,
+      "startBlocks and endBlocks must have the same length"
+    );
+    require(signerIndices.length > 0, "requires at least one signerIndex");
+
+    uint256 epochSize = getEpochSize();
+    uint256 signerIndicesIndex = 0;
+    for (uint256 i = 0; i < startBlocks.length; i = i.add(1)) {
+      if (i > 0) {
+        require(
+          startBlocks[i.sub(1)] < startBlocks[i],
+          "every startBlock must be bigger than the last one"
+        );
+        require(
+          endBlocks[i.sub(1)] >= startBlocks[i].sub(1) && endBlocks[i.sub(1)] < endBlocks[i],
+          "at least one endBlock is not in the boundaries of the next interval"
+        );
+        // is first block of Epoch
+        if (startBlocks[i].sub(1).mod(epochSize) == 0) {
+          require(
+            getValidatorFromSigner(signerIndices[signerIndicesIndex], startBlocks[i].sub(1)) ==
+              getValidatorFromSigner(signerIndices[signerIndicesIndex.add(1)], startBlocks[i]),
+            "indices do not point to the same validator"
+          );
+          signerIndicesIndex = signerIndicesIndex.add(1);
+        }
+      }
+      if (!wasDownForInterval(startBlocks[i], endBlocks[i], signerIndices[signerIndicesIndex])) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * @notice Add to the validator a new slashedInterval if the validator was not already
+   * slashed in the same epoch of the startBlock or by a previous interval that contains al 
+   * least a block of the new interval
+   */
+  function addNewSlashIntervalToValidator(address validator, uint256 startBlock, uint256 endBlock)
+    internal
+  {
+    uint256 startEpoch = getEpochNumberOfBlock(startBlock);
+    uint256 endEpoch = getEpochNumberOfBlock(endBlock);
+
+    uint256[2] memory blockInterval = slashedIntervals[validator][startEpoch];
+    require(blockInterval[0] == 0 && blockInterval[1] == 0, "already slashed in that epoch");
+
+    // Check possible crossing epoch slash from the last epoch
+    blockInterval = slashedIntervals[validator][startEpoch.sub(1)];
+    require(blockInterval[1] < startBlock, "slash shares blocks with another slash");
+
+    // Check possible crossing epoch slash from the epoch of the endEpoch
+    if (startEpoch != endEpoch) {
+      blockInterval = slashedIntervals[validator][endEpoch];
+      require(
+        blockInterval[0] == 0 || endBlock < blockInterval[0],
```

Comment:
that is not the block number in the epoch, is the block number of the chain
As the memory is initialised as zero, that's the way I could check that wasn't slashed in that "next" epoch, otherwise the second statement `endBlock < blockInterval[0]` will be false because the endBlock will be bigger that 0

I'm not checking that, from epoch to epoch-1 (`require(blockInterval[1] < startBlock, "slash shares blocks with another slash");`) because in that case, if it's zero it would be ok too